### PR TITLE
AVM: Adding bmodexp

### LINKED ISSUE #6139: New opcode: modexp
## Problem

I believe `modexp` (modular exponentiation) should be available as an opcode, as it seems like an important crypto primitive. It is also present in many other chains like EVM. While I have implemented its functionality in [Puya BigNumber](https://github.com/Helium-Labs/Puya-BigNumber), I'm afraid its cost is prohibitively expensive for most real-world applications like RSA signature verification. This is why I propose it be made an opcode.

This opcode would pave the way for several novel use cases, such as supporting RSA, which is still a popular cryptographic function seen in many places like JWT verification and DKIM in email.

## Solution

I offer to make a PR adding support for the opcode, updating several source files in `data/transactions/logic` to integrate it. Before creating the PR, I thought it best to raise an issue to provide context and see whether it's something others would like to see added. The API design would be something typical, likely `modexp(base: Bytes, exponent: Bytes, modulus: Bytes)`. The opcode cost would be a function of the input size, similar to `base64_decode`.

## Dependencies

There are no known dependencies.

## Urgency

I am aware of several developers, including myself, who wish to support RSA in some form, such as for DKIM in email or privacy-preserving JWT verification. In that sense, it seems like an important opcode.

### PR DESCRIPTION
**Pull request by mangoplane**
## Summary

Adds the new opcode `bmodexp` as described in issue #6139 to support modular exponentiation involving byte strings of up to 4096 bytes. Closes #6139

## Test Plan

- Relevant tests added to `assembler_test.go`, `evalStateful_test.go`, & `eval_test.go`
- Opcode is tested with a range of test vectors with function `TestBytesModExp`, covering panic cases, edge cases, acceptance cases and failure cases. Test vectors were generated manually or with Python.

### CODE REVIEW

File: data/transactions/logic/eval.go [hunk 1]
```diff
@@ -2325,6 +2325,33 @@ func opBytesNeq(cx *EvalContext) error {
...
+
+	z := new(big.Int).SetBytes(cx.Stack[last].Bytes)
+	y := new(big.Int).SetBytes(cx.Stack[prev].Bytes)
+	x := new(big.Int).SetBytes(cx.Stack[pprev].Bytes)
+	if z.BitLen() == 0 {
+		return errors.New("modulo by zero")
+	}
+
+	result.Exp(x, y, z)
+	if result.Sign() < 0 {
```
Comment (jannotti) [inline data/transactions/logic/eval.go:2346]: This is impossible, since we only allow positive inputs, right?

File: data/transactions/logic/eval.go [hunk 2]
```diff
@@ -2325,6 +2325,33 @@ func opBytesNeq(cx *EvalContext) error {
 	return opNot(cx)
 }
 
+func opBytesModExp(cx *EvalContext) error {
+	result := new(big.Int)
+	last := len(cx.Stack) - 1 // z
+	prev := last - 1          // y
+	pprev := last - 2         // x
+
+	if len(cx.Stack[last].Bytes) > maxStringSize || len(cx.Stack[prev].Bytes) > maxStringSize || len(cx.Stack[pprev].Bytes) > maxStringSize {
```
Comment (jannotti) [inline data/transactions/logic/eval.go:2334]: I'm doing this from phone, but `maxStringSize` is the AVM max of 4096? That's unusual for the bmath functions, they normally have a maximum of 128 bytes. I suppose you want more because RSA uses really large keys?

At any rate, it's impossible to supply inputs that are larger than this, so there's no need to check in the opcode.

File: data/transactions/logic/evalStateful_test.go [hunk 3]
```diff
@@ -3254,6 +3254,7 @@ func TestReturnTypes(t *testing.T) {
 
 		"box_create": "int 9; +; box_create",                 // make the size match the 10 in CreateBox
 		"box_put":    "byte 0x010203040506; concat; box_put", // make the 4 byte arg into a 10
+		"bmodexp":    ": byte 0x0102030405; byte 0x010203040506; byte 0x01020304050607; bmodexp",
```
Comment (jannotti) [inline data/transactions/logic/evalStateful_test.go:3257]: Was this required to make the test pass? I wouldn't have expected it to be. This test is really only a sanity check on typing, so if you did this just to exercise the opcode more, I'd rather you do that all in the opcode specific tests.

File: data/transactions/logic/eval_test.go [hunk 4]
```diff
@@ -5065,6 +5065,55 @@ func TestBitLen(t *testing.T) {
...
+
+	for _, modexpTestVector := range modexpTestVectors {
+		progText := fmt.Sprintf(`byte %s
+byte %s
+byte %s
+bmodexp
+
+byte %s
+==`, modexpTestVector.Base, modexpTestVector.Exponent, modexpTestVector.Modulus, modexpTestVector.Result)
+		if modexpTestVector.TestOutcome == Accept {
```
Comment (jannotti) [inline data/transactions/logic/eval_test.go:5106]: A `switch` on the TestOutcome might look nicer.

File: data/transactions/logic/opcodes.go [hunk 5]
```diff
@@ -81,6 +81,7 @@ const spliceVersion = 10  // box splicing/resizing
 const incentiveVersion = 11 // block fields, heartbeat
 
 const spOpcodesVersion = 11 // falcon_verify, sumhash512
+const bmodexpVersion = 11   // bmodexp
```
Comment (jannotti) [inline data/transactions/logic/opcodes.go:84]: I think we can dispense with the elaborate version constants and "experimental" stuff for this opcode. When we merge it, we're committing to get it into the next version.

**Review by jannotti — COMMENTED** [review]
This is looking quite good so far.

File: data/transactions/logic/eval.go (hunk 2 above)
Comment (jannotti) [reply data/transactions/logic/eval.go:2334]: It's ok with me, assuming we can get the cost function to properly account for long inputs. It seems like `bmodexp` wouldn't be very interesting if it can't handle common RSA key sizes.

I'd like to support bigger inputs on the other `b` opcodes too, if we can adjust cost functions appropriately.  They were first done before we could have the cost depend on size. `b+`, for example, would almost certainly be easy to adjust, `b*` might be more complex than simple length dependence. I'm somewhat worried that `bmodexp` is going to be tricky. Anyway, that should be a separate PR someday. (Do your RSA implementations require operating on RSA keys with any other operations?)

File: data/transactions/logic/evalStateful_test.go (hunk 3 above)
Comment (jannotti) [reply data/transactions/logic/evalStateful_test.go:3257]: The convention is to only add an element to that table if it's required.  For the majority of opcodes, the test conjures up "reasonable" inputs (like a number for integer inputs or a short string for byte inputs).  This table is for opcodes that need specific inputs in order to function.

File: data/transactions/logic/eval.go (hunk 1 above)
Comment (jannotti) [reply data/transactions/logic/eval.go:2346]: I'd remove it. `opBytesBin` needs to be able to handle things like `b-` going negative. `opBytesSqrt`, for example, does no similar check since it's a one-off implementation of the opcode.

File: data/transactions/logic/eval.go (hunk 2 above)
Comment (jannotti) [reply data/transactions/logic/eval.go:2334]: If it ends up being fast enough, we _could_ pick a cost that accounts for the worst case, which I suppose would three very long inputs.

Or perhaps only scales based on one of the parameters, but accounts for the worst case with the others. (I think `bmodexp` should be linear with respect to the length of the exponent, since it basically performs one operation per bit.)

File: config/config_test.go [hunk 6]
```diff
@@ -773,6 +773,7 @@ func TestLocal_ValidateP2PHybridConfig(t *testing.T) {
 
 	for i, test := range tests {
 		test := test
+		i := i
```
Comment (jannotti) [inline config/config_test.go:776]: Was lint complaining? I don't see why, unless there's code taking the address of `i` which would be unusual.

File: data/transactions/logic/doc.go [hunk 7]
```diff
@@ -257,21 +257,22 @@ var opDescByName = map[string]OpDesc{
...
+	"b*":      {"A times B. A and B are interpreted as big-endian unsigned integers.", "", nil},
+	"b<":      {"1 if A is less than B, else 0. A and B are interpreted as big-endian unsigned integers", "", nil},
+	"b>":      {"1 if A is greater than B, else 0. A and B are interpreted as big-endian unsigned integers", "", nil},
+	"b<=":     {"1 if A is less than or equal to B, else 0. A and B are interpreted as big-endian unsigned integers", "", nil},
+	"b>=":     {"1 if A is greater than or equal to B, else 0. A and B are interpreted as big-endian unsigned integers", "", nil},
+	"b==":     {"1 if A is equal to B, else 0. A and B are interpreted as big-endian unsigned integers", "", nil},
+	"b!=":     {"0 if A is equal to B, else 1. A and B are interpreted as big-endian unsigned integers", "", nil},
+	"b%":      {"A modulo B. A and B are interpreted as big-endian unsigned integers. Fail if B is zero.", "", nil},
+	"bmodexp": {"A raised to the Bth power modulo C. A, B and C are interpreted as big-endian unsigned integers limited to 4096 bytes. Fail if C is zero.", "", nil},
```
Comment (jannotti) [inline data/transactions/logic/doc.go:271]: Can you move this done to follow `bsqrt` both for consistency and because it will minimize the diff caused by autoformatting.

File: data/transactions/logic/eval.go (hunk 2 above)
Comment (jannotti) [reply data/transactions/logic/eval.go:2334]: I just read your discussion on cost from the issue more closely.  The dependence on the square of the lengths is unfortunate because it implies a custom cost function. Currently, all costs are "data directed" which is nice because it means we can create specs automatically - we can generate text that describes the cost from the constants provided in the opcode table.

I suppose we can add a way to provide both a Go function and a textual description directly in the `.go` source code. That is somewhat more fragile from the standpoint of modifications to the way we present the spec, but it doesn't see so bad.  It's probably also necessary if we ever want to support larger inputs to `b*` which I suspect is where this really coming from.

**Comment by giuliop** [comment]
`bmodexp` is useful also in different scenarios with smaller inputs so we should not penalize those cost-wise, for instance all ZKP protocols based on elliptic curves use it and they try to operate on the smallest field possible while preserving security for efficiency.

As a concrete example, smart contract verifiers for zk-proofs based on the plonk protocol generated by [AlgoPlonk](https://github.com/giuliop/AlgoPlonk) call a teal [implementation](https://github.com/giuliop/AlgoPlonk/blob/main/verifier/templateBN254.go#L364) of it 5 times, using 32-byte inputs for both curve BN254 and BLS12-381.

Considering that a BN254 verifier consumes ~145,000 opcode budget and a BLS12-381 verifier ~185,000, `bmodexp` would really help bring down that cost.

**Comment by giuliop** [comment]
> I'm doing this from phone, but `maxStringSize` is the AVM max of 4096? That's unusual for the bmath functions, they normally have a maximum of 128 bytes. I suppose you want more because RSA uses really large keys?

Isn't the maximum [64](https://github.com/algorand/go-algorand/blob/2a02530a933ad2f6bd53eddb967f6e97979d74e6/data/transactions/logic/eval.go#L53) bytes?

**Comment by jannotti** [comment]
Just to close the loop, is the suggestion that this: `x = exponent_length * max(base_length, modulus_length)**2` will work for the cost function (with appropriate scaling)?

I would be on board with that, and I'll just have to write an "escape hatch" to allow an arbitrary cost function to be written.

**Comment by algorandskiy** [comment]
Based on this [eval](https://drive.google.com/file/d/1Z6HjmV8scdztfz7ZM6f2VUNRP8nkysYd/view) some log-poly formula gives better approximation.

**Comment by giuliop** [comment]
Looks reasonable to me, the # of iterations is `exponent_length` and the amount of work performed per iteration is in first approximation modulo(base * base) so proportional to `max(base_length, modulus_length)**2`

**Comment by giuliop** [comment]
I benchmarked `bmodexp` on my own and  `exponent_length * max(base_length, modulus_length)**2` looks reasonable to me.

I benchmarked using the same byte length for all three inputs, base, mode, and exp and I get:
| Benchmark                                  | Iterations | Time (ns/op)    | Extra (extra/op) |
|--------------------------------------------|------------|-----------------|------------------|
| BenchmarkBModExp/modexp_32_bytes_inputs-12 | 78,057     | 15,019 ns/op    | 4.000 extra/op   |
| BenchmarkBModExp/modexp_64_bytes_inputs-12 | 18,218     | 66,383 ns/op    | 4.000 extra/op   |
| BenchmarkBModExp/modexp_128_bytes_inputs-12| 3,776      | 322,610 ns/op   | 4.000 extra/op   |
| BenchmarkBModExp/modexp_256_bytes_inputs-12| 556        | 2,239,315 ns/op | 4.000 extra/op   |
| BenchmarkBModExp/modexp_512_bytes_inputs-12| 63         | 18,215,421 ns/op| 4.000 extra/op   |
| BenchmarkBModExp/modexp_1024_bytes_inputs-12| 8         | 142,717,656 ns/op| 4.000 extra/op  |
| BenchmarkBModExp/modexp_2048_bytes_inputs-12| 1         | 1,147,616,375 ns/op| 4.000 extra/op |
| BenchmarkBModExp/modexp_4096_bytes_inputs-12| 1         | 9,566,156,542 ns/op| 4.000 extra/op |

If we divide the ns/op by `inputs_length**3` we get:

| Inputs Len | Cost       | Inputs Len³   | Cost / Inputs Len³ |
|-----------|------------|--------------|-------|
| 32        | 15,019     | 32,768       | 0.46   |
| 64        | 66,383     | 262,144      | 0.25   |
| 128       | 322,610    | 2,097,152    | 0.15   |
| 256       | 2,239,315  | 16,777,216   | 0.13   |
| 512       | 18,215,421 | 134,217,728  | 0.14   |
| 1024      | 142,717,656| 1,073,741,824| 0.13   |
| 2048      | 1,147,616,375| 8,589,934,592| 0.13   |
| 4096      | 9,566,156,542| 68,719,476,736| 0.14  |

Looks like the cost stabilizes after inputs length of 128 bytes.
I think we can make the cost proportional to `exponent_length * max(base_length, modulus_length)**2` and perhaps add a constant factor to account for the higher relative cost of small inputs


This is the benchmark function I'm using:
```
func BenchmarkBModExp(b *testing.B) {
	for _, byteLen := range []int{32, 64, 128, 256, 512, 1024, 2048, 4096} {
		var base, exp, mod []byte
		for _, input := range []*[]byte{&base, &exp, &mod} {
			// create random inputs of the given length, we are using math/rand without seeding,
			// so it will generate the same pseudorandom numbers for each run
			*input = make([]byte, byteLen)
			for i := range *input {
				(*input)[i] = byte(rand.Intn(256))
			}
		}
		ops := fmt.Sprintf("byte 0x%x; byte 0x%x; byte 0x%x; bmodexp; pop", base, exp, mod)
		b.Run(fmt.Sprintf("modexp_%d_bytes_inputs", byteLen), func(b *testing.B) {
			benchmarkOperation(b, "", ops, "int 1")
		})
	}
}
```

**Comment by algorandskiy** [comment]
@mangoplane please commit the benchmarks into `crypto_test.go` - that's where other cost evaluation benchmarks for crypto opcodes live.

**Comment by giuliop** [comment]
> @giuliop how about re-running modexp_1024+ (like `-count=64 I guess)` to get a better avg value?

sure, I rerun using `-benchtime=64x` to have 64 runs (except for 4096 bytes which would timeout, so I ran it 32 times) and the results are in line as before:

| Benchmark                                          | Trials | Time per op (ns/op) | Extra ops |
| -------------------------------------------------- |:------:|:------------------:|:---------:|
| BenchmarkBModExp1/modexp_1024_bytes_inputs-12      |   64   |   142,368,141       |    4.000  |
| BenchmarkBModExp1/modexp_2048_bytes_inputs-12      |   64   | 1,139,738,240       |    4.000  |
| BenchmarkBModExp1/modexp_4096_bytes_inputs-12      |   32   | 9,443,533,249       |    4.000  |

Looks to me like we can use `exponent_length * max(base_length, modulus_length)**2` and not make it more complicated than that

**Comment by giuliop** [comment]
Looking back at benchmark the 4096 byte length case for all three parameters takes 9 sec to run, so it's not feasible on the AVM unfortunately.
@jannotti what would you consider the upper limit in terms of how long an opcode can run ?

**Comment by algorandskiy** [comment]
I reproduced the results and got the log-poly model fits my data points perfectly
```
log_A coefficient: 0.0
log_B coefficient: 0.30924400959422704
log_C coefficient: 1.2483621313283038
log_A_log_B coefficient: 0.0
log_A_log_C coefficient: 0.0
log_B_log_C coefficient: 0.10928549552529052
log_A_log_B_log_C coefficient: 8.62725548783673e-05
Intercept: 0.0
R-squared (on log scale): 0.9964
Root Mean Squared Error: 98143.7211 gas units
```
image

**Comment by giuliop** [comment]
> All is not lost, RSA keys seem to be at most 4k BITS, not BYTES. Does that end up being 8^3 cheaper because of the squaring of the multiplicand length, and the factor of 8 from the exponent?
> 
> 9/512 = 0.017 which starts to become almost feasible.

Indeed the benchmarking at 512 bytes length for all parameters gives `0.018 sec`

| Inputs Len (bytes) | ns / op       | Inputs Len³   | ns / Inputs Len³ |
|-----------|------------|--------------|-------|
| 32        | 15,019     | 32,768       | 0.46   |
| 64        | 66,383     | 262,144      | 0.25   |
| 128       | 322,610    | 2,097,152    | 0.15   |
| 256       | 2,239,315  | 16,777,216   | 0.13   |
| 512       | 18,215,421 | 134,217,728  | 0.14   |

That means though that 512 bytes is still too large given that the max for a single program is `20,000 * 16 * 15ns = 0.0048 sec` , so the max length for `bmodexp`'s parameters can be 256 bytes.

This assumes we want to cap all parameters at the same size.
@mangoplane for your use case what is the max size of each of the three parameters?

If the assumption that we want to cap all parameters at the same size holds, I guess we have two options:
1) Introduce `bmodexp` for up to 64 bytes parameters for consistency with all other b-operations
2) Introduce `bmodexp` for up to 256 bytes parameters (maybe calling it `modexp`)

I would suggest to do option 1 now and then do the work to graduate all b-operations to 256 bytes parameters

**Comment by giuliop** [comment]
For the cost function I would use this formula to determine the actual opcode cost, i.e., the number of ticks:

`max(base_lenght, mod_lenght)^1.63 * exp_length / 15 + 200`

where lenght is the # of bytes.

This is how the benchmarking looks like, here I am keeping base and mod at the same byte length and varying the exp length. I am calculating the # of ticks by dividing the ns/op coming from `go test` by 15.

| Base & Mod Len | Exp Len | Average ns/op | Average 'ticks' | Formula 'ticks' |
|----------------|---------|------------|-----------------|-----------------|
| 32             | 16      | 6,445      | 430             | 503             |
| 32             | 24      | 9,274      | 618             | 654             |
| 32             | 32      | 12,665     | 844             | 806             |
| 64             | 16      | 18,432     | 1,229           | 1,138           |
| 64             | 32      | 33,163     | 2,211           | 2,076           |
| 64             | 48      | 52,550     | 3,503           | 3,013           |
| 64             | 64      | 70,248     | 4,683           | 3,951           |
| 128            | 32      | 82,731     | 5,515           | 6,005           |
| 128            | 64      | 160,046    | 10,670          | 11,810          |
| 128            | 96      | 239,196    | 15,946          | 17,615          |
| 128            | 128     | 315,648    | 21,043          | 23,420          |
| 256            | 64      | 579,067    | 38,604          | 36,135          |
| 256            | 128     | 1,166,991  | 77,799          | 72,070          |
| 256            | 192     | 1,652,136  | 110,142         | 108,006         |
| 256            | 256     | 2,121,371  | 141,425         | 143,941         |

This is the function used for the benchmarking:
```
func BenchmarkBModExp2(b *testing.B) {
    // Define the base and mod lengths, and corresponding exp lengths
    for _, byteLen := range []int{32, 64, 128, 256} {
        var expLens []int
        switch byteLen {
        case 32:
            expLens = []int{16, 24, 32}
        case 64:
            expLens = []int{16, 32, 48, 64}
        case 128:
            expLens = []int{32, 64, 96, 128}
        case 256:
            expLens = []int{64, 128, 192, 256}
        }

        for _, expLen := range expLens {
            // Generate base and mod of length byteLen
            base := make([]byte, byteLen)
            mod := make([]byte, byteLen)
            for _, input := range []*[]byte{&base, &mod} {
                for i := range *input {
                    (*input)[i] = byte(mathrand.Intn(256))
                }
            }

            // Generate exp of varying lengths
            exp := make([]byte, expLen)
            for i := range exp {
                exp[i] = byte(mathrand.Intn(256))
            }

            ops := fmt.Sprintf("byte 0x%x; byte 0x%x; byte 0x%x; bmodexp; pop", base, exp, mod)

            b.Run(fmt.Sprintf("modexp_base&mod=%d_bytes_exp=%d_bytes", byteLen, expLen), func(b *testing.B) {
                benchmarkOperation(b, "", ops, "int 1")
            })
        }
    }
}
```

**Comment by giuliop** [comment]
I think option 3 can work, developers can always check at runtime if needed the sizes of their inputs and manage in the code the cases where the opcode budget would be exceeded

File: data/transactions/logic/opcodes.go [hunk 8]
```diff
@@ -85,15 +86,41 @@ const spOpcodesVersion = 11 // falcon_verify, sumhash512
...
+// Custom cost definition for bmodexp
+func bmodExpCostFunction(stack []stackValue, depth int) int {
+	last := len(stack) - depth - 1 // mod
+	prev := last - depth - 1       // exp
+	pprev := last - depth - 2      // base
+
+	expLength := float64(len(stack[prev].Bytes))
+	modLength := float64(len(stack[last].Bytes))
+	baseLength := float64(len(stack[pprev].Bytes))
+	cost := (math.Pow(math.Max(baseLength, modLength), 1.63) * expLength / 15) + 200
```
Comment (cusma) [inline data/transactions/logic/opcodes.go:104]: Wouldn't be better to declare those parameters (`1.63`, `15` and `200`) as constants? I think it would make easier to "tune" them in one place, if necessary, instead of replacing them in all the consumers.

Maybe also some comments on how those constants have been chosen would be helpful for future readers.

Comment (cusma) [reply data/transactions/logic/opcodes.go]: Thanks for addressing this. Looks good!

**Comment by cusma** [comment]
@jannotti I was wondering if wouldn't be appropriate to have the proposed op code "cost function" parameters in the consensus configuration (instead of hardcoding them in `opcode.go` - https://github.com/algorand/go-algorand/pull/6140/commits/468210e44adbd29a6dca7acc2d342707eb8cb2b4).

**Comment by jannotti** [comment]
> @jannotti I was wondering if wouldn't be appropriate to have the proposed op code "cost function" parameters in the consensus configuration (instead of hardcoding them in `opcode.go` - [468210e](https://github.com/algorand/go-algorand/commit/468210e44adbd29a6dca7acc2d342707eb8cb2b4)).

We need to be able to tie cost functions to AVM versions, rather than consensus versions, so that existing code does not change behavior. That's why all the costs are set in `logic` package, basis on the version of bytecode running.

**Comment by algorandskiy** [comment]
@mangoplane could you merge master in?

**Comment by jannotti** [comment]
> @mangoplane could you merge master in?

I created a larger PR (#6270) addressing some related opcodes, and using the very nice custom cost function infrastructure added by @mangoplane .

I'll close this and handle merging master into that one.
